File: LinkedList/LL.js
class Node{
    constructor(d){
        this.data=d
        this.next=null;
    }
}

class LinkedList{
    constructor(){
        this.head=null
    }
    addToStart(data){
        let newData=new Node(data)
        newData.next=this.head
        this.head=newData
    }
    addAtEnd(data){
        let newData=new Node(data)
        
        //Check if LL is empty
        if(this.head==null){
            this.head=newData
        }

        //Navigate to The End of the LinkedList
        let current=this.head
        while(current.next!=null){
            current=current.next
        }
        current.next=newData
    }

    addAfterANode(data,num){
        let newData=new Node(data)

        //Check if LL is Empty
        if(this.head==null){
            console.log("Element not Found")
            return
        }

        let current=this.head
        while(current!=null && current.data!=num){
            current=current.next
        }
        if(current!=null){
            newData.next=current.next
            current.next=newData
            return
        }
        console.log("Element not Found")
    }
    
    addABeforeANode(data,num){
        let newData=new Node(data)

        //Check if LL is Empty
        if(this.head==null){
            console.log("Element not Found")
            return
        }

        let current=this.head
        while(current!=null && current.next.data!=num){
            current=current.next
        }
        if(current!=null){
            newData.next=current.next
            current.next=newData
            return
        }
        console.log("Element not Found")
    }

    deleteFromStart(){
        if(this.head==null){
            console.log("List Is Empty")
            return
        }

        this.head=this.head.next

    }

    deleteFromEnd(){
        if(this.head==null){
            console.log("List Is Empty")
            return
        }

        let current=this.head
        while(current.next.next!=null){
            current=current.next
        }

        current.next=null
    }

    printAll(){
        let current=this.head
        while(current!=null){
            console.log(current.data)
            current=current.next
        }
    }
}

let head=new LinkedList()
head.addToStart(15)
head.addToStart(10)
head.addToStart(5)
head.addAtEnd(30)

head.addAfterANode(20,15)
head.addABeforeANode(25,30)

head.printAll()

console.log("Deletion from Start")
head.deleteFromStart()
head.printAll()

console.log("Deletion from End")
head.deleteFromEnd()
head.printAll()

File: LinkedList/LL.py
class Node:
    def __init__(self,data):
        self.data=data
        self.next=None

class LinkedList:
    def __init__(self):
        self.head=None
    
    def addToStart(self,data):
        newNode=Node(data)
        newNode.next=self.head
        self.head=newNode

    def addAtEnd(self,data):
        newNode=Node(data)
        
        if self.head==None:
            print("Hii")
            newNode.next=self.head
            self.head=newNode
            return

        current=self.head
        while(current!=None):
            current=current.next
        current.next=newNode

    def printLL(self):
        while(self.head!=None):
            print(self.head.data)
            self.head=self.head.next

if __name__=='__main__':
    ll=LinkedList()
    # ll.addToStart(5)
    # ll.addToStart(10)
    ll.addAtEnd(20)
    ll.addAtEnd(40)
    ll.printLL()


